### PR TITLE
Quick search

### PR DESCRIPTION
**Pull request by justyna-13**
### Related Issues

closes #

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature
- [ ] (Next only) I've followed this [instruction](https://docs-next.vuestorefront.io/contributing/creating-changelog.html) and I've created a `.js` file with information about my Pull Request

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 1]
```diff
@@ -72,15 +117,22 @@ export default {
...
+  directives: { clickOutside },
   setup(props, { root }) {
     const { toggleCartSidebar, toggleWishlistSidebar, toggleLoginModal } = useUiState();
     const { changeSearchTerm, getFacetsFromURL } = useUiHelpers();
+    const { result, search } = useFacet();
     const { isAuthenticated, load: loadUser } = useUser();
     const { cart, load: loadCart } = useCart();
     const { load: loadWishlist } = useWishlist();
     const term = ref(getFacetsFromURL().term);
+    const searchOpen = ref(false);
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:134]:
```suggestion
    const isSearchOpen = ref(false);
```

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 2]
```diff
@@ -72,15 +117,22 @@ export default {
...
+    const searchRef = ref(null);
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:135]:
```suggestion
    const searcBarhRef = ref(null);
```

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 3]
```diff
@@ -92,6 +144,9 @@ export default {
     // TODO: https://github.com/DivanteLtd/vue-storefront/issues/4927
     const handleAccountClick = async () => {
       if (isAuthenticated.value) {
+        if (searchOpen.value) {
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:147]: we can achieve that by closing the search box after search form loses focus

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 4]
```diff
@@ -104,14 +159,66 @@ export default {
       await loadWishlist();
     });
 
+    const handleSearchInput = (input) => term.value = input;
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:162]: v-model

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 5]
```diff
@@ -104,14 +159,66 @@ export default {
...
+
+    const closeSearchAndRemoveTerm = () => {
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:164]:
```suggestion
    const closeSearch = () => {
```

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 6]
```diff
@@ -72,15 +117,22 @@ export default {
...
-    SfBadge
+    SfBadge,
+    SfSearchBar,
+    SearchResults,
+    SfOverlay
   },
+  directives: { clickOutside },
   setup(props, { root }) {
     const { toggleCartSidebar, toggleWishlistSidebar, toggleLoginModal } = useUiState();
     const { changeSearchTerm, getFacetsFromURL } = useUiHelpers();
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:128]: changeSearchTerm => `setTermForUrl`

Comment (filrak) [reply packages/core/nuxt-theme-module/theme/components/AppHeader.vue:128]: and maybe move to the search component

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 7]
```diff
@@ -1,69 +1,114 @@
...
+          </div>
+        </template>
+        <template #search>
+          <SfSearchBar
+            ref="searchRef"
+            :value="term"
+            placeholder="Search for items"
+            aria-label="Search"
+            class="sf-header__search"
+            @input="handleSearchInput"
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:63]: most likely can be just a single fn

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 8]
```diff
@@ -104,14 +159,66 @@ export default {
...
+        term.value = '';
+        return searchRef.value.$el.children[0].focus();
+      }
+    };
+
+    watch(() => term.value, (newVal, oldVal) => {
+      if (term.value.length >= 3) {
+        debounceSearchInput();
+      }
+      if (!isMobile.value && term.value.length > 0) {
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:195]: make it a single if and maybe assign to a variable
const canDoSth = (!oldVal && newVal) || (newVal.length !== oldVal.length && searchOpen.value === false && ((!isMobile.value && term.value.length > 0))

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 9]
```diff
@@ -104,14 +159,66 @@ export default {
...
+        return closeSearchAndRemoveTerm();
+      } else {
+        term.value = '';
+        return searchRef.value.$el.children[0].focus();
+      }
+    };
+
+    watch(() => term.value, (newVal, oldVal) => {
+      if (term.value.length >= 3) {
+        debounceSearchInput();
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:193]: we don't need to externalize it

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 10]
```diff
@@ -104,14 +159,66 @@ export default {
...
 
+    const handleSearchInput = (input) => term.value = input;
+
+    const closeSearchAndRemoveTerm = () => {
+      term.value = '';
+      searchOpen.value = false;
+      changeSearchTerm(term.value);
+    };
+
+    const handleSearch = async (e) => {
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:170]: just have a single debounced function that handles both search AND passing term to the url

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 11]
```diff
@@ -104,14 +159,66 @@ export default {
...
+      await search({ categorySlug: 'women-clothing-jackets', itemsPrePage: 20, filters: {} });
+    };
+
+    const debounceSearchInput = debounce(async () => {
+      await search({ categorySlug: 'women-clothing-jackets', itemsPrePage: 20, filters: {} });
+    }, 1000);
+
+    const isMobile = computed(() => mapMobileObserver().isMobile.get());
+
+    const handleCloseSearchButton = () => {
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:182]:
```suggestion
    const closeSearch = () => {
```

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 12]
```diff
@@ -1,69 +1,114 @@
...
+              >
+                <span class="sf-search-bar__icon">
+                  <SfIcon color="var(--c-text)" size="20px" icon="search" />
+                </span>
+              </SfButton>
+            </template>
+          </SfSearchBar>
+        </template>
+      </SfHeader>
+      <SearchResults :visible="searchOpen" :result="result" @closeSearchResults="closeSearchAndRemoveTerm" />
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:91]: closeSearchResults => close

**Review by filrak — CHANGES_REQUESTED** [review]

File: packages/core/nuxt-theme-module/theme/components/SearchResults.vue [hunk 13]
```diff
@@ -0,0 +1,285 @@
+<template>
+  <div>
+    <SfMegaMenu
+      :visible="toggleSearch"
+      title="Search results"
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/SearchResults.vue:5]: i18n

File: packages/core/nuxt-theme-module/theme/components/SearchResults.vue [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+      class="search"
+    >
+      <transition name="sf-fade" mode="out-in">
+        <div v-if="products.length > 0" class="search__wrapper-results" key="results">
+          <SfMegaMenuColumn title="Categories" class="sf-mega-menu-column--pined-content-on-mobile search__categories">
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/SearchResults.vue:10]: i18n

File: packages/core/nuxt-theme-module/theme/components/SearchResults.vue [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+            <SfList>
+              <SfListItem v-for="(category, key) in categories.items" :key="key">
+                <SfMenuItem :label="category.label" @click="goToPath(`/c/women/${category.slug}`)"/>
+              </SfListItem>
+            </SfList>
+          </SfMegaMenuColumn>
+          <SfMegaMenuColumn title="Product suggestions" class="sf-mega-menu-column--pined-content-on-mobile search__results">
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/SearchResults.vue:17]: i18n

File: packages/core/nuxt-theme-module/theme/components/SearchResults.vue [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+      toggleSearch.value = newVal;
+      if (toggleSearch.value) {
+        document.body.classList.add('no-scroll');
+      } else {
+        document.body.classList.remove('no-scroll');
+      }
+    });
+
+    const goToPath = (path) => {
+      emit('closeSearchResults');
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/SearchResults.vue:160]: not needed after implementing focus/close

File: packages/core/nuxt-theme-module/theme/components/SearchResults.vue [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+    visible: {
+      type: Boolean,
+      default: false
+    },
+    result: {
+      type: Object
+    }
+  },
+  setup(props, { emit, root }) {
+    const toggleSearch = ref(props.visible);
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/SearchResults.vue:146]:
```suggestion
    const isSearchOpen = ref(props.visible);
```

File: packages/core/nuxt-theme-module/theme/components/SearchResults.vue [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+                          {{ productGetters.getName(product) }}
+                        </h3>
+                      </SfButton>
+                    </template>
+                </SfProductCard>
+              </div>
+            <SfButton class="sf-button--text see-all desktop-only" @click="goToPath('/c/women/women-clothing')">See all results</SfButton>
+          </SfMegaMenuColumn>
+          <div class="action-buttons smartphone-only">
+            <SfButton class="action-buttons__button color-secondary" @click="goToPath('/c/women/women-clothing')">See all results</SfButton>
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/SearchResults.vue:94]: i18n

File: packages/core/nuxt-theme-module/theme/components/SearchResults.vue [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+                        @click="goToPath(`/p/${productGetters.getId(product)}/${productGetters.getSlug(product)}`)"
+                      >
+                        <h3 class="sf-product-card__title">
+                          {{ productGetters.getName(product) }}
+                        </h3>
+                      </SfButton>
+                    </template>
+                </SfProductCard>
+              </div>
+            <SfButton class="sf-button--text see-all desktop-only" @click="goToPath('/c/women/women-clothing')">See all results</SfButton>
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/components/SearchResults.vue:91]: i18n

**Review by filrak — COMMENTED** [review]
- get rid of hardcoded links/data

File: packages/core/nuxt-theme-module/theme/components/AppHeader.vue [hunk 20]
```diff
@@ -104,14 +155,60 @@ export default {
...
+    };
+
+    const handleSearch = debounce(async (paramValue) => {
+      if (!paramValue.target) {
+        term.value = paramValue;
+      } else {
+        term.value = paramValue.target.value;
+      }
+      setTermForUrl(term.value);
+      getSearchTermFromUrl(term.value);
```
Comment (justyna-13) [inline packages/core/nuxt-theme-module/theme/components/AppHeader.vue:171]:
```suggestion
```

**Pull request merged**